File: verify/parity.js
// Parity fixture generator: same matrix as snap.zig / parity.py.
// Emits full logical cells (glyph:color:emphasis) plus the ASCII render,
// so parity covers the complete cell contract, not just glyphs.
const path = require('path');
const R = require(path.join(__dirname, '..', 'rail.js'));

const states = ['idle', 'listening', 'captured', 'thinking', 'speaking', 'acting',
  'waiting', 'needs_input', 'complete', 'interrupted', 'error'];
const widths = [25, 37, 49, 61];
const ticks = [0, 1, 2, 3, 4, 7, 11, 26, 53];
// 64-bit boundary seeds: zero, >2^31 (32-bit truncation trap), 2^53-1
// (Number precision edge), 2^64-1 (full range; BigInt required in JS)
const seeds = [0n, 2147483649n, 9007199254740991n, 18446744073709551615n];

const lines = [];
function cellsStr(cells) {
  return cells.map((c) => `${c.glyph}:${c.color}:${c.emphasis}`).join(',');
}
function snapLine(state, name, w, tick, moff, prog, seed) {
  const q = tick % 5;
  const cells = R.frame(state, {
    width: w, tick, entryTick: 0, inputQ: q, outputQ: q, seed,
    progress: prog, motion: moff ? 'off' : 'normal',
  });
  lines.push(`${name}|${w}|${tick}|${moff ? 1 : 0}|${seed}|${R.renderAscii(cells)}|${cellsStr(cells)}`);
}

for (const st of states) {
  for (const w of widths) {
    for (const tick of ticks) {
      for (const moff of [false, true]) {
        snapLine(st, st, w, tick, moff, null, 9n);
        if (st === 'acting') snapLine(st, 'acting_det', w, tick, moff, tick / 64, 9n);
      }
    }
  }
}
// seed-boundary block: seeded states only
for (const seed of seeds) {
  for (const st of ['thinking', 'waiting', 'error']) {
    for (const tick of [4, 26]) {
      snapLine(st, st, 49, tick, false, null, seed);
    }
  }
}
process.stdout.write(lines.join('\n') + '\n');

File: verify/parity.py
"""Parity fixture generator: same matrix as snap.zig / parity.js, via rail.py.
Emits full logical cells (glyph:color:emphasis) plus the ASCII render."""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
import rail  # noqa: E402  (the Signal Rail engine one directory up)

states = ["idle", "listening", "captured", "thinking", "speaking", "acting",
          "waiting", "needs_input", "complete", "interrupted", "error"]
widths = [25, 37, 49, 61]
ticks = [0, 1, 2, 3, 4, 7, 11, 26, 53]
# 64-bit boundary seeds (see parity.js for rationale)
seeds = [0, 2147483649, 9007199254740991, 18446744073709551615]

lines = []


def cells_str(cells):
    return ",".join(f"{c.glyph}:{c.color}:{c.emphasis}" for c in cells)


def snap_line(state, name, w, tick, moff, prog, seed):
    q = tick % 5
    cells = rail.frame(state, width=w, tick=tick, entry_tick=0, input_q=q,
                       output_q=q, seed=seed, progress=prog,
                       motion="off" if moff else "normal")
    lines.append(f"{name}|{w}|{tick}|{1 if moff else 0}|{seed}|"
                 f"{rail.render_ascii(cells)}|{cells_str(cells)}")


for st in states:
    for w in widths:
        for tick in ticks:
            for moff in (False, True):
                snap_line(st, st, w, tick, moff, None, 9)
                if st == "acting":
                    snap_line(st, "acting_det", w, tick, moff, tick / 64, 9)

for seed in seeds:
    for st in ("thinking", "waiting", "error"):
        for tick in (4, 26):
            snap_line(st, st, 49, tick, False, None, seed)

sys.stdout.write("\n".join(lines) + "\n")
